secure/storage/src/storage.rs

```diff
@@ -10,68 +14,310 @@ use libra_crypto::{
 /// This is the Libra interface into secure storage. Any storage engine implementing this trait
 /// should support both key/value operations (e.g., get, set and create) and cryptographic key
 /// operations (e.g., generate_key, sign_message and rotate_key).
-pub trait Storage: KVStorage + CryptoStorage {}
-
-impl<T> Storage for T where T: KVStorage + CryptoStorage {}
 
-/// This is a hack that allows us to convert from SecureBackend into a useable T: Storage, since
-/// Box<dyn Storage> is not T: Storage.
-pub struct BoxStorage(pub Box<dyn Storage>);
+/// This is a hack that allows us to convert from SecureBackend into a useable
+/// T: Storage. This boilerplate can be 100% generated by a proc macro.
+/// TODO(fga): generate this boilerplate with a proc macro.
```

Comment:
In this precise case, I feel that we could go a long way with regular macros (*).

(*) We may have to go 2nd-order, though:
https://www.reddit.com/r/rust/comments/dopgqi/higher_order_macros/
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=51143f2e22e1635a0361310aaa21f4ac

Reply:
more precisely, enum_dispatch?

Reply:
+1 to update the comment